### PR TITLE
feat(channels): add private channel option

### LINKED ISSUE #425: Feature Request: Open private channels
**Is your feature request related to a problem? Please describe.**
For testing purposes it would be awesome to easily create private channels in polar

**Describe the solution you'd like**
When opening a channel in polar it would be cool if there was a simple checkbox named "private" which should be unchecked by default.
Private channel lines could be displayed semitransparent in the view graph.

**Additional context**
I am testing stuff where I need to know if routes over private channels are found.

### PR DESCRIPTION
**Pull request by bolatovumar**
Closes #425

### Description

Added option to create a private channel. Private channels show up as low-opacity lines connecting nodes.

### Steps to Test

1. Create a test network
2. Open "Create a channel" modal
3. Notice there is now a "Make channel private" option
4. Check the aforementioned option
5. Notice that a private channel is created which shows up as a low-opacity line as opposed to a full-opacity line for regular channels

### Screenshots

Screen Shot 2021-02-21 at 6 25 17 PM

### CODE REVIEW

File: src/components/designer/custom/Link.tsx
Comment (bolatovumar) [inline src/components/designer/custom/Link.tsx:91]: Create a lower-opacity line here if channel is private.

File: src/components/designer/lightning/actions/OpenChannelModal.tsx
Comment (bolatovumar) [inline src/components/designer/lightning/actions/OpenChannelModal.tsx:63]: Replaced `any` with an inline type.

File: src/components/designer/lightning/actions/OpenChannelModal.tsx
Comment (bolatovumar) [inline src/components/designer/lightning/actions/OpenChannelModal.tsx:80]: Notice the new `isPrivate` option here.

File: src/components/designer/lightning/actions/OpenChannelModal.tsx
Comment (bolatovumar) [inline src/components/designer/lightning/actions/OpenChannelModal.tsx:141]: New checkbox to make the channel private.

File: src/components/designer/link/Channel.tsx
Comment (bolatovumar) [inline src/components/designer/link/Channel.tsx:46]: I'm not sure if showing a `true` or `false` string here is the best. Maybe showing `Yes` or `No` would be better?

File: src/i18n/locales/de-DE.json
Comment (bolatovumar) [inline src/i18n/locales/de-DE.json:107]: Not sure what to do with non-English locales.

File: src/lib/lightning/clightning/clightningService.ts
Comment (bolatovumar) [inline src/lib/lightning/clightning/clightningService.ts:78]: Notice that I named the field `isPrivate` instead of a simpler `private` because `private` is a reserved keyword in JavaScript so I wanted to avoid it.

File: src/lib/lightning/clightning/clightningService.ts
Comment (bolatovumar) [inline src/lib/lightning/clightning/clightningService.ts:111]: Instead of adding yet another parameter to the function signature I replaced it with a single `options` parameter which contains all options which will be used to open a channel.

File: src/lib/lightning/clightning/clightningService.ts
Comment (bolatovumar) [inline src/lib/lightning/clightning/clightningService.ts:127]: `c-lightning` uses the `announce` field to mark a channel as private. However, when you retrieve channels from `c-lightning` the channel is not marked as private (`private` field is always `false`).

File: src/lib/lightning/eclair/eclairService.spec.ts
Comment (bolatovumar) [inline src/lib/lightning/eclair/eclairService.spec.ts:76]: Added this because I updated the type from `any` to a more specific one so I had to add the missing property to the test. Channel flag `1` in `eclair` is for public channels.

File: src/lib/lightning/eclair/eclairService.ts
Comment (bolatovumar) [inline src/lib/lightning/eclair/eclairService.ts:84]: Previously we were passing `capacity` as a number here when we expect a `string`.

File: src/lib/lightning/eclair/eclairService.ts
Comment (bolatovumar) [inline src/lib/lightning/eclair/eclairService.ts:88]: Channel flag `0` is for private channels in `eclair`.

File: src/lib/lightning/eclair/types.ts
Comment (bolatovumar) [inline src/lib/lightning/eclair/types.ts:68]: Update the type from `any` to something more specific which include the properties we actually use.

File: src/lib/lightning/eclair/types.ts
Comment (bolatovumar) [inline src/lib/lightning/eclair/types.ts:83]: Updated the type to a more specific one as the only possible values here are `0` and `1`.

File: src/lib/lightning/lnd/mappers.ts
Comment (bolatovumar) [inline src/lib/lightning/lnd/mappers.ts:29]: Since `private` field doesn't exist on `PendingChannel` type so I'm setting it to `false` here. I guess I could also make the field nullable instead and assign it to `null` here?

File: src/utils/chart.ts
Comment (bolatovumar) [inline src/utils/chart.ts:174]: Just moved this into its own constant so I can specify the type and added the `isPrivate` field.

**Comment by jamaljsr** [comment]
Hey @bolatovumar thanks for working on this. I'll work on reviewing this later today 👍

File: src/i18n/locales/de-DE.json
Comment (jamaljsr) [reply src/i18n/locales/de-DE.json:107]: You only need to update the `en-US.json` file. All of the other language files will be re-generated and overwritten by Crowdin.

Also, I'd prefer to capitalize the `P` to match the rest of the title case labels displayed in the sidebar.

File: src/lib/lightning/clightning/clightningService.ts
Comment (jamaljsr) [reply src/lib/lightning/clightning/clightningService.ts:78]: Thanks for the rationale 👍

File: src/lib/lightning/clightning/clightningService.ts
Comment (jamaljsr) [reply src/lib/lightning/clightning/clightningService.ts:111]: Good refactor. Thanks 🥇

File: src/lib/lightning/clightning/clightningService.ts
Comment (jamaljsr) [reply src/lib/lightning/clightning/clightningService.ts:127]: Do you know if the root cause of this problem is with the `c-lightning-REST` API or with `lightningd`? We should open an issue in their repos if the RPC isn't working as expected.

I don't think it's a good idea to display the "Make private" checkbox and then show a non-private channel after it's created. That makes it seem like Polar is broken.

File: src/lib/lightning/eclair/eclairService.spec.ts
Comment (jamaljsr) [reply src/lib/lightning/eclair/eclairService.spec.ts:76]: Great improvement. I should have done this initially instead of using `any`. 🥇

File: src/lib/lightning/eclair/eclairService.ts
Comment (jamaljsr) [reply src/lib/lightning/eclair/eclairService.ts:84]: Good catch!

File: src/lib/lightning/lnd/mappers.ts
Comment (jamaljsr) [reply src/lib/lightning/lnd/mappers.ts:29]: Using `false` is fine with me here.

**Review by jamaljsr — CHANGES_REQUESTED** [review]
Excellent PR @bolatovumar. The code looks really really good. Thank you for the refactoring improvements.

I just have two requests blocking approval:
1. The private channels for `c-lightning` nodes are not working correctly. I left a comment below on this. 
2. When you run `yarn test:ci`, you'll see there are still some conditionals not covered by test cases. I think you just need to add a couple more tests to cover the opposite case for `isPrivate`.

Also, just a suggestion in general, not a deal-breaker for this PR. It would be better to split up the changes into multiple smaller commits. This one is pretty large. I know it's easier to do it all in one go, but the smaller commits are much easier to review and result in a cleaner git history.

File: src/components/designer/link/Channel.tsx
Comment (bolatovumar) [reply src/components/designer/link/Channel.tsx:46]: Ok, I will leave it as is then.

File: src/lib/lightning/clightning/clightningService.ts
Comment (bolatovumar) [reply src/lib/lightning/clightning/clightningService.ts:127]: I don't know where the issue lies. I will have to look deeper into it. I believe I am using the right field here yet the channel is not private no matter what value I send it.

> I don't think it's a good idea to display the "Make private" checkbox and then show a non-private channel after it's created. That makes it seem like Polar is broken.
Yeah, I agree. I just wanted to create a PR anyway to start the discussion and figure out what's wrong in the process.

File: src/lib/lightning/lnd/mappers.ts
Comment (bolatovumar) [reply src/lib/lightning/lnd/mappers.ts:29]: Cool.

File: src/components/designer/custom/Link.tsx
Comment (bolatovumar) [reply src/components/designer/custom/Link.tsx]: Done here: 60b888bf568a58ebe5e4de517a76157e27883d96.

File: src/i18n/locales/de-DE.json
Comment (bolatovumar) [reply src/i18n/locales/de-DE.json]: Updated capitalization here: 05b62b2a35c15538a0831ca2398b07a2c1de6d30

**Comment by bolatovumar** [comment]
> 1. The private channels for `c-lightning` nodes are not working correctly. I left a comment below on this.
Yeah, I will look into this.

> 2. When you run `yarn test:ci`, you'll see there are still some conditionals not covered by test cases. I think you just need to add a couple more tests to cover the opposite case for `isPrivate`.
Will do.

> Also, just a suggestion in general, not a deal-breaker for this PR. It would be better to split up the changes into multiple smaller commits. This one is pretty large. I know it's easier to do it all in one go, but the smaller commits are much easier to review and result in a cleaner git history.
I assumed that the commits would be squashed into a single one at merge time for some reason so just did one giant commit. I would have split it into a bunch of smaller commits otherwise.

File: src/components/designer/custom/Link.tsx
Comment (jamaljsr) [reply src/components/designer/custom/Link.tsx]: Hey, sorry to nitpick, but what I meant was to return `opacity` in the existing `useMemo` call that's already present on line 50. Since this block is already converting `link.properties` values into component vars, it makes sense to determine the `opacity` in there as well.

Something like this is what I had in mind:
```js
  const { leftStop, rightStop, leftColor, rightColor, opacity } = useMemo(() => {
    ...
    // the opacity of the link's path defaults to 100%
    let opacity = 1;
    if (link.properties) {
      const {
        type,
        direction,
        toBalance,
        capacity,
        isPrivate,
      } = link.properties as LinkProperties;
      if (type === 'open-channel') {
        ...
      }
      // use a lighter opacity for private channels
      if (isPrivate) opacity = 0.5;
    }
    return {
      leftStop,
      rightStop,
      leftColor,
      rightColor,
      opacity,
    };
  }, [link.properties, theme]);
```

File: src/lib/lightning/clightning/clightningService.ts
Comment (jamaljsr) [reply src/lib/lightning/clightning/clightningService.ts:127]: Ok cool, let me know if you'd like me to help troubleshoot in any way.

**Comment by bolatovumar** [comment]
@jamaljsr I think I found where the bug lies with `c-lightning`. It's in the REST API implementation on this line: https://github.com/Ride-The-Lightning/c-lightning-REST/blob/22bb6c34960d09b8ec3c834a606b863446a3ef4f/controllers/channel.js#L77
Since this line double-bangs the string input for the `announce` field if it's `'0'` or `'false'` the result will always be `true` as double-bang of any non-empty string will be true, so the channel is never set to private. I will file the issue in that repo.

**Comment by bolatovumar** [comment]
Opened an issue here: https://github.com/Ride-The-Lightning/c-lightning-REST/issues/55#issue-819545464

**Comment by jamaljsr** [comment]
@bolatovumar Good find! 👍  

I'd like to keep this PR blocked until that issue is resolved in `c-lightning-REST` repo. I'll try to keep an eye on it, but feel free to update me here when that one gets resolved.

**Comment by bolatovumar** [comment]
@jamaljsr I created a PR to fix the bug in the `c-lightning-REST` repo but I was wondering how I could test if it would actually fix the issue before that PR is merged. I tried changing this line: https://github.com/jamaljsr/polar/blob/f9a88ce022c9b2633209b31a5f6fee10518486cb/docker/clightning/Dockerfile#L137 to point to the tip of my for of the `c-lightning-REST` repo which has the fix but it doesn't seem to cut it. Do I need to build `c-lightning` from source using this Dockerfile or something?

**Comment by jamaljsr** [comment]
That should work. Just edit the `Dockerfile`, then build the image. Polar should just use the new image. Did you use the same label `polarlightning/clightning:0.9.3`? Maybe try deleting the existing image before rebuilding. Also try adding `console.log`s to your fork repo to confirm Polar is using the new image.

**Comment by bolatovumar** [comment]
> That should work. Just edit the `Dockerfile`, then build the image. Polar should just use the new image. Did you use the same label `polarlightning/clightning:0.9.3`? Maybe try deleting the existing image before rebuilding. Also try adding `console.log`s to your fork repo to confirm Polar is using the new image.

Ok, I build the Docker image while pointing to my fork of the `c-lightning-REST` repo which has this fix and after waiting for a while for it to build on my laptop it worked. See proof of screenshot:

Screen Shot 2021-03-05 at 10 16 56 PM

Will this fix get applied automatically once my PR in `c-lightning-REST` is merged in or will the images have to be re-built manually for all versions of `c-lightning`?

**Comment by jamaljsr** [comment]
Thanks for the update. Lookin good 👍 

> Will this fix get applied automatically once my PR in `c-lightning-REST` is merged in or will the images have to be re-built manually for all versions of `c-lightning`?

Yes, the images would need to be rebuilt and pushed to DockerHub. Unfortunately, even after I do that, anyone who has already pulled the images from DockerHub will not automatically get the update. Only users who have never pulled the `c-lightning` image would receive the updated one when the network is started.

This is the first time where we've needed to update existing images. I'll have to put some thought into developing a strategy to handle this now as well as going forward. 

Off the top of my head, I have two ideas:
1. Create new image labels containing a suffix (ex: `polarlightning/clightning:0.9.3-1`). Then update the [node.json](https://github.com/jamaljsr/polar/blob/master/docker/nodes.json) file to include this suffix. We would just need to be mindful to maintain backwards compatibility with older versions of Polar.
2. Add a button on the **Manage Images** screen to remove local `polarlightning/*` images. The next time a network is started, it will pull the updated images. The downside is that this is not very obvious to users. They would not assume to do this when the `c-lightning` private channels are not displaying correctly.

If you have any feedback or other suggestions, please do share.

**Comment by bolatovumar** [comment]
> 1. Create new image labels containing a suffix (ex: `polarlightning/clightning:0.9.3-1`). Then update the [node.json](https://github.com/jamaljsr/polar/blob/master/docker/nodes.json) file to include this suffix. We would just need to be mindful to maintain backwards compatibility with older versions of Polar.

In this case previous versions of `c-lightning` will continue working but will not get this fix, right? If that's the case I wonder if it makes sense to add a warning of some sort in the UI if a user is trying to open a private channel using an older version of the `c-lightning` image. It's a bit cumbersome but could be helpful.

> 2. Add a button on the **Manage Images** screen to remove local `polarlightning/*` images. The next time a network is started, it will pull the updated images. The downside is that this is not very obvious to users. They would not assume to do this when the `c-lightning` private channels are not displaying correctly.

This might still be useful even if it's not obvious as there could be other use cases for this if a user wants to remove local images for some other unrelated reason in the future.

**Comment by bolatovumar** [comment]
Related `c-lightning-REST` PR merged: https://github.com/Ride-The-Lightning/c-lightning-REST/pull/56

**Comment by jamaljsr** [comment]
Thanks for the update. I'll update the images soon.

Regarding the UI changes. I like your idea about showing the warning when opening a private channel. We could use the docker image hash to determine if the user has the new image or an old one pulled locally. If they have the old one, instruct them to click the button on the **Manage Images** screen. This seems like a reasonable solution. Thanks for the suggestion 👍

**Review by jaonoctus — APPROVED** [review]

**Comment by jamaljsr** [comment]
I rebased this PR and made a couple fixes and improvements. Thanks again for adding this feature. Apologies for taking so long to merge.

**Pull request merged**
